compiler/src/main/kotlin/dev/zacsweers/metro/compiler/ir/transformers/DependencyGraphTransformer.kt

```diff
@@ -2352,6 +2304,61 @@ internal class DependencyGraphTransformer(
     }
   }
 
+  private fun checkScopingCompatibility(
+    node: DependencyGraphNode,
+    binding: Binding,
+    bindingStack: IrBindingStack,
+  ) {
+    val bindingScope = binding.scope ?: return
+
+    if (node.scopes.isEmpty() || !node.hasAccessToScope(bindingScope)) {
```

Comment:
This is the notable logic change, the rest of the method should be effectively the same, just pulled into a helper.

This used to check if the bindingScope was in the node's scopes, and now it checks the node's scopes and its parents' scopes